### ISSUE TITLE
Next steps research plan [Feb 2]

### ISSUE THREAD

**Issue by LiamConnell** [LiamConnell/nanochat#7]
Read up on the experiments I was pursuing recently and issue #1  and write out what my next steps should be as a comment.

**Comment by LiamConnell** [comment]
## Next Steps Research Plan (Feb 2, 2026)

Based on the experiment logs from January 28-29 and issue #1 (Contextual Distillation research plan), here are the recommended next steps:

---

### 🔴 Critical Blocker: Model Memory Issues

The d32 model (1.5B params) requires too much memory for training even on A100 40GB. The last GCP job (8586008974711586816) showed:
- Model loads fine (~6.8GB)
- Baseline evaluations work (MMLU: 0.3906, ARC-Easy: 0.6641)
- **OOM on first training iteration** - optimizer states + gradients push memory over 40GB

**Options to address this (in order of recommendation):**

1. **Create a d20 SFT checkpoint** (~561M params)
   - Would fit on T4 (16GB) or L4 (24GB) GPUs
   - Much faster iteration cycles
   - Lower cost (~$100 for full run vs ~$1000 for d32)
   - Need to run SFT training first or download pre-trained d20

2. **Enable gradient checkpointing**
   - Trade compute for memory
   - Would need to implement in the GPT model class

3. **Use memory-efficient optimizers**
   - 8-bit Adam (bitsandbytes)
   - Adafactor with memory optimization

4. **Multi-GPU setup**
   - 2x A100 with distributed training
   - Higher cost but would work

---

### ✅ What's Working

These components have been validated and are ready for use:

| Component | Status | Notes |
|-----------|--------|-------|
| QA Generation Pipeline | ✅ Working | Gemini 3 Flash Preview, generates high-quality pairs |
| KRS Evaluation Framework | ✅ Working | exact_match, F1, forgetting ratio all validated |
| GCP Infrastructure | ✅ Working | Container builds, GCS upload/download, job submission |
| Checkpoint Patching | ✅ Working | Handles missing keys, device placement fixed |
| Baseline Evaluation | ✅ Working | MMLU, ARC-Easy evals run successfully |

---

### 📋 Recommended Action Plan

**Phase 1: Resolve Memory Blocker**
- [ ] Option A: Create d20 SFT checkpoint by training from base model
- [ ] Option B: Implement gradient checkpointing in GPT model
- [ ] Verify training loop completes without OOM on 1 GPU

**Phase 2: Run Distillation Smoke Test**
- [ ] Submit successful distillation job (even if just 3 iterations)
- [ ] Verify checkpoint saving works
- [ ] Verify QA evaluation runs post-training

**Phase 3: Full Contextual Distillation Experiment (per Issue #1)**
- [ ] Select a target document/corpus to internalize
- [ ] Generate comprehensive QA pairs (100-500+)
- [ ] Run full training with proper hyperparameters
- [ ] Evaluate: KRS on held-out QA, baseline benchmarks for forgetting

**Phase 4: Analysis & Iteration**
- [ ] Compare: distilled model vs RAG vs fine-tuning baselines
- [ ] Analyze forgetting metrics
- [ ] Experiment with different question generation strategies

---

### 💡 Quick Wins

1. **Lower the batch size** - Try `device-batch-size: 1` with gradient accumulation (already attempted, may need more tuning)

2. **Disable optimizer state sharding** - Check if Muon optimizer has memory-saving modes

3. **Use `PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True`** - Already in experiment notes, ensures better memory allocation

---

### Open Questions

1. Should we prioritize training a d20 model first, or push to get d32 working?
2. For the full contextual distillation experiment, what document/corpus should we use as the target?
3. Should we track experiments in W&B or keep logs local for now?

---

*Automated by Claude Code*

**Comment by LiamConnell** [comment]
Lets go with the d20. Is there a huggingface checkpoint that we can use instead of training it from scratch?

**Comment by LiamConnell** [comment]
## D20 HuggingFace Checkpoint Research

Great news! There are several HuggingFace checkpoints available for the d20 model (~561M params) that can be used instead of training from scratch.

---

### 🎯 Recommended: jasonacox's d20 Checkpoints

**jasonacox** has published the complete d20 training pipeline on HuggingFace in native nanochat checkpoint format:

| Model | Stage | HuggingFace Repo | Notes |
|-------|-------|------------------|-------|
| **nanochat-d20-561M-pretrain** | Base | [jasonacox/nanochat-d20-561M-pretrain](https://huggingface.co/jasonacox/nanochat-d20-561M-pretrain) | Step 021400 |
| **nanochat-560M-midtrain** | Midtrain | [jasonacox/nanochat-560M-midtrain](https://huggingface.co/jasonacox/nanochat-560M-midtrain) | - |
| **nanochat-560M-sft** | SFT | [jasonacox/nanochat-560M-sft](https://huggingface.co/jasonacox/nanochat-560M-sft) | Step 000700, ready for distillation |

**Key advantages:**
- ✅ Native nanochat `.pt` format (not HuggingFace format)
- ✅ Compatible with our `checkpoint_manager.py` and `build_model()`
- ✅ All training stages available (pretrain → midtrain → SFT)
- ✅ MIT licensed

---

### 📥 Download Instructions

```python
from huggingface_hub import snapshot_download
import os

# Download SFT checkpoint (recommended for contextual distillation)
model_path = snapshot_download(
    repo_id="jasonacox/nanochat-560M-sft",
    cache_dir=os.path.expanduser("~/.cache/nanochat/hf_downloads")
)

# Or download base pretrain checkpoint
model_path = snapshot_download(
    repo_id="jasonacox/nanochat-d20-561M-pretrain",
    cache_dir=os.path.expanduser("~/.cache/nanochat/hf_downloads")
)
```

Then use with nanochat:
```python
from nanochat.checkpoint_manager import build_model
import glob

checkpoint_files = glob.glob(os.path.join(model_path, "model_*.pt"))
step = int(os.path.basename(checkpoint_files[0]).split("_")[-1].split(".")[0])
model, tokenizer, meta_data = build_model(model_path, step, device, phase="eval")
```

---

### 📊 Alternative Options

| Provider | Model | Params | Format | Notes |
|----------|-------|--------|--------|-------|
| nanochat-students | [base-d20](https://huggingface.co/nanochat-students/base-d20) | 561M | HuggingFace | Requires `trust_remote_code=True` |
| sdobson | [nanochat](https://huggingface.co/sdobson/nanochat) | 561M | Native + HF | Post-SFT, has web demo |

---

### 🔧 Compatibility with Checkpoint Patching

Our `checkpoint_manager.py` already has `_patch_missing_keys()` to handle architecture differences:
- `bigram_lambdas` → initialized to zeros (disabled)
- `value_embeds.*` → initialized to zeros
- `ve_gate.weight` → initialized to zeros

This means **older d20 checkpoints should work** even if they don't have the newer architecture features. The patching code will initialize missing parameters with sensible defaults.

---

### ✅ Recommended Next Steps

1. **Download jasonacox's SFT checkpoint:**
   ```bash
   python -c "from huggingface_hub import snapshot_download; snapshot_download('jasonacox/nanochat-560M-sft', cache_dir='$HOME/.cache/nanochat/chatsft_checkpoints/d20')"
   ```

2. **Verify it loads correctly:**
   ```bash
   uv run python -c "
   from nanochat.checkpoint_manager import load_model
   model, tok, meta = load_model('sft', device='cpu', phase='eval', model_tag='d20')
   print(f'Model loaded: {sum(p.numel() for p in model.parameters())/1e6:.1f}M params')
   "
   ```

3. **Run distillation smoke test with d20 instead of d32**

This should resolve the memory issues since d20 (561M) requires ~2.2GB vs d32 (1.5B) requiring ~6GB, plus optimizer states.

---

*Automated by Claude Code*

**Comment by LiamConnell** [comment]
ok run a distillation smoke-test with the d20 model.

**Comment by LiamConnell** [comment]
kick it off as an experiment with gh experiment tracking in issues

**Comment by LiamConnell** [comment]
## D20 Distillation Smoke Test Submitted

I've set up and submitted the d20 distillation smoke test as requested.

### What Was Done

1. **Downloaded d20 SFT checkpoint** from [jasonacox/nanochat-560M-sft](https://huggingface.co/jasonacox/nanochat-560M-sft) (~561M params)
2. **Verified model loads correctly** - the checkpoint manager patches missing keys automatically
3. **Updated `distillation_smoketest.yaml`** to use d20 instead of d32:
   - Changed from A100 (40GB) to L4 (24GB) - sufficient for the smaller model
   - Updated model-tag from d32 to d20
   - Increased batch size from 1 to 2 (fits on L4 with d20)
4. **Uploaded d20 checkpoint to GCS**: `gs://adk-rlm-1769034783-nanochat-test/chatsft_checkpoints/d20/`
5. **Submitted VertexAI job**: `9205370471707574272`

### Experiment Tracking

Created experiment tracking issue: **#11** (with `Experiment` + `Running` labels)

### Monitoring

```bash
# Check status
gcloud ai custom-jobs describe 9205370471707574272 --region=us-central1 --project=adk-rlm-1769034783

# View logs
gcloud logging read 'resource.type="ml_job" AND resource.labels.job_id="9205370471707574272"' --limit=50 --project=adk-rlm-1769034783
```

### Console URL
https://console.cloud.google.com/ai/platform/locations/us-central1/training/9205370471707574272?project=381631550294

---
*Automated by Claude Code*
